python_modules/dagster/dagster/core/definitions/assets_info.py

```diff
@@ -64,6 +67,188 @@ def __new__(
         )
 
 
+def _build_graph_dependencies(
+    graph_def: GraphDefinition,
+    parent_handle: NodeHandle,
+    outputs_by_graph_handle: Mapping[NodeHandle, Mapping[str, Tuple[str, NodeHandle]]],
+    non_asset_inputs_by_node_handle: Dict[NodeHandle, Sequence[Tuple[str, NodeHandle]]],
+    assets_defs_by_node_handle,
+):
+    """
+    Scans through every node in the graph, making a recursive call when a node is a graph.
+
+    Builds two dictionaries:
+
+    outputs_by_graph_handle: A mapping of every graph node handle to a dictionary with each out
+        name as a key and a tuple containing (op output name, op node handle) as the value
+
+    non_asset_inputs_by_node_handle: A mapping of all node handles to all upstream node handles
+        that are not assets. Each key is a node handle with tuple value (node_output_name, output_node)
+    """
+    dep_struct = graph_def.dependency_structure
+    for sub_node_name, sub_node in graph_def.node_dict.items():
+        curr_node_handle = NodeHandle(sub_node_name, parent=parent_handle)
+        if sub_node.is_graph:
+            _build_graph_dependencies(
+                sub_node.definition,
+                curr_node_handle,
+                outputs_by_graph_handle,
+                non_asset_inputs_by_node_handle,
+                assets_defs_by_node_handle,
+            )
+            outputs_by_graph_handle[curr_node_handle] = {
+                mapping.definition.name: (
+                    mapping.maps_from.output_name,
+                    NodeHandle(mapping.maps_from.solid_name, parent=curr_node_handle),
+                )
+                for mapping in sub_node.definition.output_mappings
+            }
+        non_asset_inputs_by_node_handle[curr_node_handle] = [
+            (
+                output_handle.output_def.name,
+                NodeHandle(output_handle.solid_name, parent=parent_handle),
+            )
+            for output_handle in dep_struct.all_upstream_outputs_from_solid(sub_node_name)
+            if NodeHandle(output_handle.solid.name, parent=parent_handle)
+            not in assets_defs_by_node_handle
+        ]
+
+
+def _get_dependency_node_handles(
+    non_asset_inputs_by_node_handle: Mapping[NodeHandle, Sequence[Tuple[str, NodeHandle]]],
+    outputs_by_graph_handle: Mapping[NodeHandle, Mapping[str, Tuple[str, NodeHandle]]],
+    dep_node_handles_by_node: Mapping[NodeHandle, Sequence[NodeHandle]],
+    curr_node_handle: NodeHandle,
+    graph_output_name: Optional[str] = None,
+):
```

Comment:
A return type annotation would be helpful here.

Reply:
Thanks, added!